stl/inc/generator

```diff
@@ -362,7 +362,7 @@ public:
         return *this;
     }
 
-    _NODISCARD _Ref operator*() const noexcept {
+    _NODISCARD _Ref operator*() const noexcept(is_nothrow_copy_constructible_v<_Ref>) {

```

Comment:
This is constructing, but IIRC not necessarily copy constructing, a `_Ref`. Should this be `is_nothrow_constructible_v<_Ref, decltype(*_Coro.promise()._Top.promise()._Ptr)>`?

Reply:
I think `is_nothrow_copy_constructible_v` is correct.

- `_Ptr` has type `add_pointer_t<_Yielded>`.
- `_Yielded` is either `_Ref` (if it is a reference) or `const _Ref&`.
- so `decltype(*_Ptr)` is either `_Ref&` or `const _Ref&`.
- If `_Ref` is a reference, then the construction from `_Ref&` is non-throwing, and `is_nothrow_copy_constructible_v<_Ref>` happens to be `true` as well.
- If `_Ref` is not a reference, then `decltype(*_Ptr)` must be `const _Ref&`, and the condition is equivalent to `is_nothrow_copy_constructible_v<_Ref>`.

The standard [also says `noexcept(is_nothrow_copy_constructible_v<reference>)`](https://eel.is/c++draft/coro.generator.iterator#lib:operator*,generator::iterator).

Reply:
I think the condition needs to be `noexcept(static_cast<_Ref>(*_Coro.promise()._Top.promise()._Ptr))`.

If `_Ref` is an rvalue reference, this is a cast from lvalue to rvalue reference. Such a cast is non-throwing IIUC, but `is_nothrow_constructible_v` is `false`.

The standard says [`noexcept(is_nothrow_copy_constructible_v<reference>)`](https://eel.is/c++draft/coro.generator.iterator#lib:operator*,generator::iterator), which also needs to be fixed.

Reply:
On further thought, I believe the logic here is "If `_Ref` is a reference type, this function doesn't throw, which is consistent with `is_nothrow_copy_constructible_v<_Ref>`. If it's not a reference type, it must be a (cv-unqualified?) object type, and `_Ptr` is a pointer to (const?) `_Ref` so the `static_cast` expression in fact does copy a `_Ref`."

What we've merged here is correct, but we should nonetheless change it in a followup to obviously agree with the Standard wording.